core/src/main/java/io/questdb/griffin/engine/table/PrometheusMetricsRecordCursorFactory.java

```diff
@@ -0,0 +1,364 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.table;
+
+import io.questdb.Metrics;
+import io.questdb.cairo.AbstractRecordCursorFactory;
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.DataUnavailableException;
+import io.questdb.cairo.GenericRecordMetadata;
+import io.questdb.cairo.TableColumnMetadata;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.RecordCursor;
+import io.questdb.cairo.sql.RecordMetadata;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.metrics.MetricsRegistry;
+import io.questdb.metrics.PrometheusFormatUtils;
+import io.questdb.metrics.Target;
+import io.questdb.std.LongList;
+import io.questdb.std.Misc;
+import io.questdb.std.str.DirectUtf8Sink;
+import io.questdb.std.str.DirectUtf8String;
+import io.questdb.std.str.Utf8Sequence;
+import io.questdb.std.str.Utf8String;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+import static io.questdb.metrics.MemoryTagLongGauge.MEMORY_TAG_PREFIX;
+
+public final class PrometheusMetricsRecordCursorFactory extends AbstractRecordCursorFactory {
+    public static final RecordMetadata METADATA;
+    private final PrometheusMetricsCursor cursor;
+    private final PrometheusMetricsRecord record;
+
+    public PrometheusMetricsRecordCursorFactory(CairoConfiguration configuration) throws SqlException {
+        super(METADATA);
+        if (!configuration.getMetrics().isEnabled()) {
+            throw SqlException.$(0, "metrics are disabled! try setting `metrics.enabled=true` in `server.conf`");
+        }
+
+        DirectUtf8Sink sink = new DirectUtf8Sink(255);
+        LongList values = new LongList(10);
+        DirectUtf8String value = new DirectUtf8String();
+        record = new PrometheusMetricsRecord(sink, values, value);
```

Comment:
this is a little messy: confused object ownership - three objects above can be created by the record where they are used